crates/cargo-contract/src/cmd/verify.rs

```diff
@@ -62,9 +72,100 @@ impl VerifyCommand {
+        if let Some(path) = &self.contract {
+            self.verify_contract(manifest_path, verbosity, path)
+        } else if let Some(path) = &self.wasm {
+            self.verify_wasm(manifest_path, verbosity, path)
+        } else {
+            anyhow::bail!("Either --wasm or --contract must be specified")
+        }
+    }
+
+    /// Verify `.wasm` binary.
+    fn verify_wasm(
+        &self,
+        manifest_path: ManifestPath,
+        verbosity: Verbosity,
+        path: &PathBuf,
+    ) -> Result<VerificationResult> {
+        // 1. Read code hash binary from the path.
+        let file = File::open(path)
+            .context(format!("Failed to open contract binary {}", path.display()))?;
+
+        let mut ref_reader = BufReader::new(file);
+        let mut ref_buffer = Vec::new();
+        ref_reader
+            .read_to_end(&mut ref_buffer)
+            .context(format!("Failed to read contract binary {}", path.display()))?;
+
+        let reference_code_hash = code_hash(&ref_buffer);
+
+        // 2. Call `cargo contract build` in the release mode.
+        let args = ExecuteArgs {
+            manifest_path: manifest_path.clone(),
+            verbosity,
+            optimization_passes: Some(contract_build::OptimizationPasses::Z),
+            build_mode: BuildMode::Release,
+            build_artifact: BuildArtifacts::CodeOnly,
+            extra_lints: false,
+            ..Default::default()
+        };
+
+        let build_result = execute(args)?;
+
+        // 4. Grab the code hash from the built contract and compare it with the reference
+        //    one.
+        let built_wasm_path = if let Some(m) = build_result.dest_wasm {
+            m
+        } else {
+            // Since we're building the contract ourselves this should always be
+            // populated, but we'll bail out here just in case.
+            anyhow::bail!("\nThe workspace contract does not contain a Wasm binary,\n\
+                therefore we are unable to verify the contract."
+                .to_string()
+                .bright_yellow())
+        };
+
+        let file = File::open(&built_wasm_path).context(format!(
+            "Failed to open contract binary {}",
+            &built_wasm_path.display()
+        ))?;
 
+        let mut target_reader = BufReader::new(file);
+        let mut target_buffer = Vec::new();
+        target_reader
+            .read_to_end(&mut target_buffer)
+            .context(format!("Failed to read contract binary {}", path.display()))?;
+
+        let output_code_hash = code_hash(&target_buffer);
+
+        if output_code_hash != reference_code_hash {
+            anyhow::bail!(format!(
+                "\nFailed to verify the authenticity of wasm binary at {} against the workspace \n\
+                found at {}.",
+                format!("`{}`", path.display()).bright_white(),
+                format!("{:?}", manifest_path.as_ref()).bright_white()).bright_red()
+            );
+        }
```

Comment:
If we are generating and comparing the hashes, instead of just the raw files, we should at least display them here.